### DISCUSSION TITLE
supabase-auth-helpers invalidates session on refresh and defaults to anonkey

### DISCUSSION THREAD

**Discussion by perhp** [supabase/supabase#6080] [Questions] [answered]
I'm facing a weird issue where I've had a policy for a long time which seemingly has stopped working.

This is the policy:
`(EXISTS ( SELECT users.id, users.role FROM users WHERE ((users.id = uid()) AND ((users.role)::text = 'admin'::text))))`

Also I get this error and not the "policy violated" error. But if I turn off policies, I get no errors at all.
`{
    "message":"JSON object requested, multiple (or no) rows returned",
    "details":"Results contain 0 rows, application/vnd.pgrst.object+json requires 1 row"
}`

Now if I disable the policy, everything seems to be working fine. So I'm also wondering how come I'm not getting the "violating rls policy" error?

EDIT:
Seems like it's not just this policy, I also have a `(uid() = id)` failing - sometimes only tho ???

EDIT:
After some investigation it seems like my auth are getting invalidated whenever i refresh the browser. If I look at the request headers, it will send the access token (jwt) until i refresh, then it just sends the anonkey in authorization? Is this expected behaviour? I'm using @supabase/supabase-js to handle requests to db and  @supabase/supabase-auth-helpers to handle my auth. Also this only seems to be a clientside issue, as my serverside code doesn't have any issue authenticating and authorizing the user no matter how many times i refresh (getServersideProps with withAuthRequired), although the clientside will not set the correct headers when the page is rendered..

Its a lot like this issue https://github.com/supabase/supabase/issues/5852#issuecomment-1079633416, although I'm getting it both locally and in production (not using cypress, just all the time pretty much).
I haven't had this exact issue until yesterday, but I'm unable to locate if there's any changes in libraries I'm using that could cause this.

**Comment by burggraf** [comment]
I always use `auth.uid()` in my security policies.  It used to be required to use the `auth.` schema part here, but I'm not sure if this is necessary today.

**Reply by perhp** [reply]
I tried but unfortunately didn't seem to do it. Also my problem seems to be unrelated to my policies, and more related to the clientside auth handling, as it will just stop setting the access token in my authorization headers, and use the anonkey instead. Just like described in this issue #5852 , although im getting it locally and live.. It seems to be happening whenever I refresh the browser.

I use @supabase/supabase-js to handle requests to db and @supabase/supabase-auth-helpers to handle my auth..

**Comment by GaryAustin1** [comment]
Is supabase.js storing user info in localstorage?  On refresh, by default settings in client, it uses localstorage to get the user session info.
Not sure what would have changed recently though.

**Reply by perhp** [reply]
It seems to me like its stored in cookies and not localstorage. What would be the expected behaviour?

**Reply by GaryAustin1** [reply]
What is your supabase createClient call?

**Reply by perhp** [reply]
@supabase/supabase-auth-helpers is handling creation of my supabaseClient, which seems to be working fine. It's created with whatever credentials i have in NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY or something like that.

**Reply by GaryAustin1** [reply]
Sorry, interpreted your initial comment that you did have client side specific code.  I notice auth-helpers has some changes 8 days ago and 5 days ago, but not familiar with it.

**Reply by perhp** [reply]
I have this usercontext provider, it will probably be a good showcase for this.

```
import { supabaseClient } from "@supabase/supabase-auth-helpers/nextjs";
import { useUser } from "@supabase/supabase-auth-helpers/react";
import { createContext, useEffect, useState } from "react";

export const UserContext = createContext(null);

export const UserProvider = ({ children }) => {
  const { user: authUser } = useUser();
  const [user, setUser] = useState(null);

  useEffect(() => {
    console.log(authUser);
    if (!authUser) {
      return;
    }

    (async () => {
      const { data: user, error } = await supabaseClient
        .from("users")
        .select()
        .eq("id", authUser.id)
        .single();
      if (error) {
        return;
      }

      setUser(user);
    })();
  }, [authUser]);

  return <UserContext.Provider value={user}>{children}</UserContext.Provider>;
};
```

So I'm just waiting for an authUser from useUser, and whenever i get something from there that's not null, I'll use the supabaseClient to fetch the logged in user from the db. Now the request just after i have received my authUser, does not always set the access token, it will just set the supabase anon key.

![image](https://user-images.githubusercontent.com/20334004/160245660-321d915d-84c6-4442-8520-5531517b37ae.png)

So as shown in the image:
First i get a null, I dont do anything. Then I get an authuser, and i try to request the user, and it will fail since my authorization header will look like this:

![image](https://user-images.githubusercontent.com/20334004/160245060-393048bc-e8cd-4903-ac2f-d4fd299d83ca.png)


Now the next interesting bit is, it doesn't always fail:

![image](https://user-images.githubusercontent.com/20334004/160245092-ce622009-d0f5-4ec4-931d-cf15e1bdb711.png)

all these request are exact same as the failed one, just working this time, and has a correct JWT access token in the authorization header

**Reply by perhp** [reply]
> Sorry, interpreted your initial comment that you did have client side specific code. I notice auth-helpers has some changes 8 days ago and 5 days ago, but not familiar with it.

Yes, the guy in the other issue had luck downgrading @supabase/supabase-js, but unfortunately didn't help for me, I'm pretty lost at what to try next..

**Reply by GaryAustin1** [reply]
Seems like a timing issue.
At a minimum you should open a new discussion with auth-helpers in the name, or open an issue there.  Not going to get attention of someone familiar with that code with the the current title.

**Reply by perhp** [reply]
Thats very true, I'll fix that

**Reply by perhp** [reply]
https://github.com/supabase/supabase-js/blob/45079cf97af503f2c8d3a8fd9658df29b03ede4c/src/SupabaseClient.ts#L266

It seems  `@supabase/supabase-js` doesn't agree I have a session, so it defaults to `this.supabaseKey`, I'll go ahead and create a issue over there to further investigate what I'm doing wrong..

**Reply by fanclubfm** [reply]
```
return createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    { autoRefreshToken: false, persistSession: false }
  );
```
Not sure if this matters, but while looking at using supabase-auth-helpers I noticed that it creates supabase client with { autoRefreshToken: false, persistSession: false } .  Probably because it has it's own logic on refreshing tokens and cookies.

**Reply by perhp** [reply]
after some investigation i think the issue is somewhere in gotrue-js, as downgrading that resolves my issues

**Comment by perhp** [comment, marked as answer]
For any bypassers: forcing @supabase/gotrue-js to version 1.20.0 fixes my issue

**Reply by perhp** [reply]
Add in your package.json
```
For yarn
"resolutions": {
    "@supabase/gotrue-js": "1.20.0"
}
```

```
For npm
"overrides": {
    "@supabase/gotrue-js": "1.20.0"
}
```

Although this "fix" should probably be used with care

**Comment by psteinroe** [comment]
We are experiencing exactly the same issue (easily reproducible by using a context similar to the one above). However, a downgrade  for `gotrue-js` did not fix it. We also had to override the `supabase-js` version.

```
for pnpm
"pnpm": {
  "overrides": {
    "@supabase/gotrue-js": "1.20.0",
    "@supabase/supabase-js": "1.27.0"
  }
}
```

**Comment by psteinroe** [comment]
Probably related issues:
- https://github.com/supabase-community/supabase-auth-helpers/issues/52
- https://github.com/supabase/supabase/issues/5852

**Comment by thorwebdev** [comment]
Will dig into this more detailed tmr, but just to make sure, are you importing supabase from the auth helpers `@supabase/supabase-auth-helpers/nextjs` as mentioned in the docs: https://github.com/supabase-community/supabase-auth-helpers/blob/main/src/nextjs/README.md#client-side-data-fetching-with-rls ?

**Reply by psteinroe** [reply]
awesome, thanks! Yes, we are only using the client from the package. FYI, there are a few similar issues that point to a regression in gotrue-js, meaning that the problem probably is not with supabase-auth-helpers but with gotrue. I listed the related issues and discussions in an issue there: https://github.com/supabase/gotrue-js/issues/267

**Comment by thorwebdev** [comment]
I'm not able to reproduce this with `gotrue-js:1.22.12`. Could you please create a repo to reproduce? This is the code I'm using: https://github.com/vercel/nextjs-subscription-payments/blob/main/utils/useUser.tsx#L43-L59 and it's loading my user data correctly with the user's token.

Can I quickly confirm that everyone is making sure to only fetch data client side after the user is set: https://github.com/supabase-community/supabase-auth-helpers/blob/next/src/nextjs/README.md#client-side-data-fetching-with-rls

```js
 useEffect(() => {
    async function loadData() {
      const { data } = await supabaseClient.from('test').select('*');
      setData(data);
    }
    // Only run query once user is logged in.
    if (user) loadData();
  }, [user]);
```
